lib/phoenix_component.ex

```diff
@@ -290,6 +290,139 @@ defmodule Phoenix.Component do
     quote do
       import Phoenix.LiveView
       import Phoenix.LiveView.Helpers
+      import unquote(__MODULE__), only: [attr: 2, attr: 3]
+
+      Module.register_attribute(__MODULE__, :__attrs__, accumulate: true)
+      Module.register_attribute(__MODULE__, :__components_calls__, accumulate: true)
+      Module.put_attribute(__MODULE__, :__components__, %{})
+
+      @on_definition {unquote(__MODULE__), :__on_definition__}
+      @before_compile {unquote(__MODULE__), :__before_compile__}
+    end
+  end
+
+  @doc "Defines an attribute for the component"
+  defmacro attr(name, type, opts \\ []) do
+    quote bind_quoted: [
+            name: name,
+            type: type,
+            opts: opts,
+            line: __CALLER__.line
+          ] do
+      Phoenix.Component.validate_attr!(name, type, opts, line, __ENV__.file)
+      Module.put_attribute(__MODULE__, :__attrs__, %{name: name, type: type, opts: opts, line: line})
+    end
+  end
+
+  @doc false
+  def validate_attr!(name, type, opts, line, file) do
+    validate_attr_type!(name, type, line, file)
+    validate_attr_opts!(name, opts, line, file)
+  end
+
+  defp validate_attr_type!(name, type, line, file) do
+    if type != :any do
+      message = """
+      invalid type `#{inspect(type)}` for attr `#{inspect(name)}`. \
+      Currently, only type `:any` is supported.\
+      """
+      raise CompileError, line: line, file: file, description: message
```

Comment:
Please rename all of those errors from `CompileError` to `ArgumentError`. The stacktrace will naturally have the correct file and line.